src/server/operators/process.rs

```diff
@@ -0,0 +1,212 @@
+        let and2 = Conditions::And(Box::new(Conditions::Leaf(b.clone())), Box::new(Conditions::Leaf(c.clone())));
+        let res = Conditions::Or(Box::new(and1), Box::new(and2));
+
+        assert_eq!(res, dnf_helper(cond));
+    }
+
+    #[test]
+    fn test_two_and_or() {
+        let a = Condition { lhs: Type::LabelKey(String::from("A")), rhs: Type::LabelValue(String::from("A")), op: Op::Eq };
+        let b = Condition { lhs: Type::LabelKey(String::from("B")), rhs: Type::LabelValue(String::from("B")), op: Op::Eq };
+        let c = Condition { lhs: Type::LabelKey(String::from("C")), rhs: Type::LabelValue(String::from("C")), op: Op::Eq };
+        let d = Condition { lhs: Type::LabelKey(String::from("D")), rhs: Type::LabelValue(String::from("D")), op: Op::Eq };
+
+
+        let or1 = Conditions::Or(Box::new(Conditions::Leaf(a.clone())), Box::new(Conditions::Leaf(b.clone())));
+        let or2 = Conditions::Or(Box::new(Conditions::Leaf(c.clone())), Box::new(Conditions::Leaf(d.clone())));
+        let cond = Conditions::And(Box::new(or1), Box::new(or2));
+
+        let and1 = Conditions::And(Box::new(Conditions::Leaf(c.clone())), Box::new(Conditions::Leaf(a.clone())));
+        let and2 = Conditions::And(Box::new(Conditions::Leaf(d.clone())), Box::new(Conditions::Leaf(a.clone())));
+        let and3 = Conditions::And(Box::new(Conditions::Leaf(c.clone())), Box::new(Conditions::Leaf(b.clone())));
+        let and4 = Conditions::And(Box::new(Conditions::Leaf(d.clone())), Box::new(Conditions::Leaf(b.clone())));
+        let res = Conditions::Or(Box::new(Conditions::Or(Box::new(and1), Box::new(and2))), Box::new(Conditions::Or(Box::new(and3), Box::new(and4))));
+
+        assert_eq!(res, dnf_helper(cond));
+    }
+
+    #[test]
+    fn test_deep_or() {
+        let a = Condition { lhs: Type::LabelKey(String::from("A")), rhs: Type::LabelValue(String::from("A")), op: Op::Eq };
+        let b = Condition { lhs: Type::LabelKey(String::from("B")), rhs: Type::LabelValue(String::from("B")), op: Op::Eq };
+        let c = Condition { lhs: Type::LabelKey(String::from("C")), rhs: Type::LabelValue(String::from("C")), op: Op::Eq };
+        let d = Condition { lhs: Type::LabelKey(String::from("D")), rhs: Type::LabelValue(String::from("D")), op: Op::Eq };
+        let e = Condition { lhs: Type::LabelKey(String::from("E")), rhs: Type::LabelValue(String::from("E")), op: Op::Eq };
+        let f = Condition { lhs: Type::LabelKey(String::from("F")), rhs: Type::LabelValue(String::from("F")), op: Op::Eq };
+
+        let or1 = Conditions::Or(Box::new(Conditions::Leaf(a)), Box::new(Conditions::Leaf(b)));
+        let and2 = Conditions::And(Box::new(Conditions::Leaf(c)), Box::new(Conditions::Leaf(d)));
+        let and3 = Conditions::And(Box::new(Conditions::Leaf(e)), Box::new(Conditions::Leaf(f)));
+
+        let and12 = Conditions::And(Box::new(or1), Box::new(and2));
+        let and123 = Conditions::And(Box::new(and12), Box::new(and3));
+        // TODO: Test this??
```

Comment:
Regarding query efficiency on top of query correctness, I think this test (or something of this form) would be a good example for further optimisations.

We note that the final form of this DNF would be `(a AND c AND d AND e AND f) OR (b AND c AND d AND e AND f)`. Rather than computing the result of each block of ANDs separately during query evaluation, there should be wins for getting the result of `x = (c AND d AND e AND f)`, then simply performing `(a AND x) OR (b AND x)` over the result.

Perhaps somewhere in the dnf conversion we can take note of the AND-blocks being distributed over the ORs to take note of these overlapping computation.

Not a priority, making a note of this in #28

Reply:
Additionally, I think putting off the equality assertion is fine for now. We'll need a helper function to test for logical equality rather than structural equality, since the structure would change whenever we update the dnf conversion algorithm.